[android] ANDROID_STL=c++_shared

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40176 [android] gradle version update
* #39587 [android] test_app example linking to pytorch_android aar content
* **#39588 [android] ANDROID_STL=c++_shared**
* #39507 [android] Add libtorch headers to pytorch_android aar
* #39584 [android][cmake] Remove NO_EXPORT for libtorch mobile build

Before this diff we used c++_static linking.
Users will dynamically link to libpytorch_jni.so and have at least one more their own shared library that probably uses stl library.

We must have not more than one stl per app. ( https://developer.android.com/ndk/guides/cpp-support#one_stl_per_app )

To have only one stl per app changing ANDROID_STL way to  c++_shared, that will add libc++_shared.so to packaging.

Differential Revision: [D22118031](https://our.internmc.facebook.com/intern/diff/D22118031)

IvanKobzarev committed:
[WIP][android] ANDROID_STL=c++_shared

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP][android] ANDROID_STL=c++_shared"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] ANDROID_STL=c++_shared"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] ANDROID_STL=c++_shared"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] ANDROID_STL=c++_shared"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] ANDROID_STL=c++_shared"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] ANDROID_STL=c++_shared"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] ANDROID_STL=c++_shared"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] ANDROID_STL=c++_shared"




[ghstack-poisoned]

ljk53 commented on scripts/build_pytorch_android.sh:
```diff
@@ -59,9 +59,18 @@ INCLUDE_DIR=$PYTORCH_ANDROID_DIR/pytorch_android/src/main/cpp/libtorch_include
 mkdir -p $INCLUDE_DIR
 
-for abi in $(echo $ABIS_LIST | tr ',' '\n')
+IFS=', ' read -r -a abis_array <<< "$ABIS_LIST"
+for abi in arm64-v8a armeabi-v7a x86_64 x86
+do
+  if [[ ! ${abis_array[*]} =~ $abi ]]
+  then
+    rm -f $LIB_DIR/$abi
+    rm -f $INCLUDE_DIR/$abi
+  fi
+done
```
I think this is to address broken symbolic links left from previous build, right? I just updated this logic recently - I simply cleared $LIB_DIR / $INCLUDE_DIR at each rebuild as these directories only contain symbolic links (super cheap to recreate).

IvanKobzarev committed:
Update on "[android] ANDROID_STL=c++_shared"


Before this diff we used c++_static linking.
Users will dynamically link to libpytorch_jni.so and have at least one more their own shared library that probably uses stl library.

We must have not more than one stl per app. ( https://developer.android.com/ndk/guides/cpp-support#one_stl_per_app )

To have only one stl per app changing ANDROID_STL way to  c++_shared, that will add libc++_shared.so to packaging.







[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] ANDROID_STL=c++_shared"


Before this diff we used c++_static linking.
Users will dynamically link to libpytorch_jni.so and have at least one more their own shared library that probably uses stl library.

We must have not more than one stl per app. ( https://developer.android.com/ndk/guides/cpp-support#one_stl_per_app )

To have only one stl per app changing ANDROID_STL way to  c++_shared, that will add libc++_shared.so to packaging.







[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] ANDROID_STL=c++_shared"


Before this diff we used c++_static linking.
Users will dynamically link to libpytorch_jni.so and have at least one more their own shared library that probably uses stl library.

We must have not more than one stl per app. ( https://developer.android.com/ndk/guides/cpp-support#one_stl_per_app )

To have only one stl per app changing ANDROID_STL way to  c++_shared, that will add libc++_shared.so to packaging.







[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] ANDROID_STL=c++_shared"


Before this diff we used c++_static linking.
Users will dynamically link to libpytorch_jni.so and have at least one more their own shared library that probably uses stl library.

We must have not more than one stl per app. ( https://developer.android.com/ndk/guides/cpp-support#one_stl_per_app )

To have only one stl per app changing ANDROID_STL way to  c++_shared, that will add libc++_shared.so to packaging.

Differential Revision: [D22118031](https://our.internmc.facebook.com/intern/diff/D22118031)

[ghstack-poisoned]
